### ISSUE TITLE
Dependency versioning

### ISSUE THREAD

**Issue by eloycoto** [cgrates/cgrates#179]
Hi,

In Foehn we are having problems with cgrates dependencies. As normal we freeze in one commit and build our new virtualservers based on the version that we want. When we need to update we re-sync cgrates repo.  

One of the problems that we have it's that sometimes the dependencies are updating and the current project is always looking for the last commit on master, so our version can't compile. 

I think that project need to move to GoDeps. It's the standard in go projects and I think that it's easy to use, plus is easy for new devs etc.. 

This is a small howto:
https://blog.codeship.com/godep-dependency-management-in-golang/

In the other hand, GoDep is used by a lot of successful projects: kubernetes, packetbeat, etc.. 

If you are agree on that maybe I can spend time on this. Quite important for us. 
Regards.

**Comment by rif** [comment]
Considering the new 1.5 vendor experiment I think we should go this way with dependency versioning.

godep added some support for this https://github.com/tools/godep#go-15-vendor-experiment but I think they suggest to add the vendor folder to the vcs and I rather not do that.

There is this tool https://github.com/Masterminds/glide that allows version pinning and management for vendor folder in a transparent way. Until go tools will add support for vendor folder management glide might be the best choice.

Here is an article introducing glide: http://engineeredweb.com/blog/2015/go-1.5-vendor-handling/

What do you think about this?

**Comment by eloycoto** [comment]
Hi rif, 

From my POV Glide has a small problem, that doesn't have reference to specific commits. So if you want to use an old version (3 weeks old) and you make a change in kamevapi project in master, the old version will not compile and this broke "our" versions. 

Does this make sense for you?
Regards.

**Comment by rif** [comment]
You can put a specific commit using the ref attribute.

This example is using master but it can be a specific commit: https://github.com/Masterminds/glide#glideyaml

The intent is to freeze the revision using `glide pin` command when making new CGRateS versions.

Did I understand your issue, correctly?

**Comment by rif** [comment]
I guess this is all glide.yaml support at the moment: https://github.com/Masterminds/glide/blob/master/docs/example-glide.yaml

**Comment by eloycoto** [comment]
Hi, 

Looks good, if the commit version/version can be used it'll great. 

Regards.

**Comment by eloycoto** [comment]
Hi Rif, 

With that glide.yml version are not referred to commits or version so it will not work if any dependency changes and not compile a code from XX weeks ago. 

I mean, you try to compile the code from 0.9.6 version that you have in download section and it will not compile if kamevapi for example has a problem in master branch. So this .yml didn't fix the issue at all, but I need to say that it's easy to install/list dependencies. 

Regards.

**Comment by rif** [comment]
My intent is to leave master glide.yaml without references. It is not recommended to use master in production.
With every new release that will make we will use glide pin command to freeze the refs we have tested before release (for that release only), and will keep master without refs after the release was made.

After we hit 1.0 stable release we plan to use something like http://nvie.com/posts/a-successful-git-branching-model/ and every supported release branch will have a different glide.yml.

**Comment by eloycoto** [comment]
Ok, but now the version 0.9.6 it's not starting in Debian 8. So the stable release need to be patched or have a different branch for bug fixes.

In the other hand, when the 1.0 version is planning? From 0.9.5 to 0.9.6 was a year.. from April danb recommended to be in master branch. 

Regards.

**Comment by rif** [comment]
The current plan is to make releases more often. Hopefully we will will make another release next week and after that there is only one extra release before 1.0.

As we did not use glide for 0.9.6 the best would be to add a glide.yaml and put some refs to few packages to make it compile.

DanB recommended master as we did not have a recent release and master is usually stable enough. There is an exception this week as we merged the ha branch and we had not enough time to test it.

We do our best to hit stable asap and after that we will have long supported releases and migration scripts for backward incompatible changes.

Bottom line is that for next week even master should become stable enough and the packages should become a simple build script run again. Sorry for the current instability and we appreciate you understanding and support (really! :).

**Comment by eloycoto** [comment]
Ok, 

And how the release will be supported? When released a new version it's not more supported? You need to update always? 

About the releases, for example if you update the releases and you don't have a different branch maybe we need to update to a new release that implies database changes, config files changes etc.. if it's a security bug it's kinda crazy for a production environment.  

My POV as a user, you should release version 1.0 provide support of bugfixes for a year, and release a new 1.X version every half a year. Every year release a new version with one year of support. 

I'm waiting for the new version.. and we want to freeze on that version for a while.

**Comment by danbogos** [comment]
Hey Eloy,

We have discussed things internally considering your ideas also and here is how we will go forward:

 * Master is considered the developer way to go as stable. Of course we still expect more user testing on it before entering into production but it should be fairly stable for testing/PoC environments. There are always features getting into which can disturb some functionality which might have been missed by us but since we will run more often complete local and call tests it should be close to the stability of the packaged versions. What happened this week was the exception since we did merge a serious branch (hapool) into master and in the last 2 days I was not able to find the time for the changes I needed to do. Will work on improving this part in the future. If you check history, we were pretty stable till last 2 days.
 * Regarding dependencies, we will keep them as specific commits for each package we depend on in the glide.yaml. They will be updated regularly so we don't get behind too much (although this requires some efforts from us, but will do it for the sake of the master stability). If you plan changing the commit references on your side, you can always keep a separate instance of glide.yaml and overwrite ours when you build your own packages.
 * We plan releasing rc7 starting of next week (need to finish aliasing and solve all sort of minor tasks for this).
 * We plan releasing rc8 as soon as Diameter support and High-Availability are completed.
 * We plan releasing the most wanted 0.9.1 (first stable release) sometimes end of the year/starting of next one. After that point will branch 0.9 which will become LTS (one year) and in that branch will push just bug fixes, no more features. Each bug fixes release will get into 0.9 branch (eg: 0.9.2, 0.9.3 and so on).
 * After freezing 0.9 branch for features will start 1.0 (first release 1.0.1) and continue with 1.1 (first release 1.1.1) and so on. 

As always, we are open to take suggestions/comments on all these.

DanB

**Comment by eloycoto** [comment]
Dear Dan: 

About the version, what about [git hooks](https://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks)? I think that at the moment it's the best way: 

- Every time that you checkout any commit you have the your glide.yml with your ref. 
- If you need a new version of any packet, you will make `glide up`. 
- When you commit you don't need to do anything. Git hook will run `git pin` and write the glide.yml and commit that file in the same commit. 

This solve: 

- Different branch always have the specific versions that need to work with.  
- Each commit will have the right dependencies. 
- You don't need to do any extra work on releases it's automatic. 
- You only need to do glide rebuild when fetch changes from another repo. 

What do you think? is it valid for you?

**Comment by danbogos** [comment]
Eloy,

By using specific commits in the glide.yaml you get exactly this freezing of specific versions since glide.yaml is archived by git. As developer we will need to checkout master of each dependency and decide what we "pin" into glide.yaml. Branches will have their own glide.yaml so don't need to do any additional checks (eventually update there if upstream packages provide bug fixing).
Regarding automatic binding, I am not sure we want to do this since there were cases in the past also when we needed to use versions from upstreams which were different from their master due to bugs on their side. 

Does it make sense to you?

DanB

**Comment by eloycoto** [comment]
Hi DanB: 

A few examples:  

**CASE 1**
You're using *cgrates/rpclient* as dependency, see what happens: 
 
1. (9:00 AM) You are working with the revision 1 of rpcclient and all works, and your glide.yml looks like this: 
```
package: github.com/cgrates/cgrates
import:
  - package: github.com/cgrates/rpcclient
```
2. At 10:00, another dev make a big change in *rpcclient* project, and you're not aware of that but this version 2 broke cgrates compilation. 

3. So a new user of Cgrates that it's on USA, in different timezone want to build the master version, but he can't compile.. The reason dependencies are not versioning.

If this glide  file was as the following, the USA customer will have a stable version tested by you:
```
package: github.com/cgrates/cgrates
import:
  - package: github.com/cgrates/rpcclient
    ref: 1
```
When you want to update rpcclient dep to another version ,you only need to do `glide up -u` and references will be updated, when you want to commit glide pin will be executed directly from your git hook. From my point of view this is easy and tested 

**CASE 2**:
You're working in branch *HA_RATER* and you need to make changes in kamevapi module too, that are in the branch *HA_RATER*, so your glide file can be the following:
```
package: github.com/cgrates/cgrates
import:
  - package: github.com/cgrates/rpcclient
    ref: @HA_RATER
```

If I want to checkout *HA_RATER* branch, it'll work like a charm. Again this will processes by `glide` and pin for the branch will be automatic. 

**CASE 3**

In Foehn we want to have our fork, that it's 3 weeks old, if this timeframe you change your deps we can't compile old version, the dependencies are not versioning. We need to update and sometimes this involve database changes, config files changes, etc.. 


Finally, you want to have the latest master as default, but that doesn't mean that will work in another dev servers, Travis or another CI servers, etc 

As I explain, this way is how works the large projects of golang, as example: [kubenetes](https://github.com/kubernetes/kubernetes) with 17K commits and more than 500 contributors, and they work like this from the start and it's was too easy to all contributors to help the project. 

Regards.

**Comment by danbogos** [comment]
For the sake of archiving, we have agreed with Eloy that the glide versioning which we use today is the right way to go further.

DanB

**Comment by eloycoto** [comment]
Hi @danbogos,

Sorry, my week was incredible busy. Yes, all looks correct from the commit 5229e616f8e3081cfb8f68ce06f6ef42f917d1ef 

Many thanks again.
